agent/fw_wordpress.c

```diff
@@ -21,156 +22,73 @@
-  nr_buffer_add(buf, NR_PSTR("\0"));
-
-  /*
-   * If the first flag is still set, no valid theme paths existed and we
-   * should abort.
-   */
-  if (first) {
-    nrl_verbosedebug(NRL_FRAMEWORK, "%s: no valid elements in the path array",
-                     __func__);
-  } else {
-    regex = nr_regex_create(nr_buffer_cptr(buf), NR_REGEX_CASELESS, 1);
-  }
-
-  nr_buffer_destroy(&buf);
-  return regex;
-}
-
-static nr_regex_t* compile_regex_for_constant(const char* constant,
-                                              const char* suffix TSRMLS_DC) {
-  nr_regex_t* regex = NULL;
-  zval* value = nr_php_get_constant(constant TSRMLS_CC);
+static nr_matcher_t* create_matcher_for_constant(const char* constant,
+                                                 const char* suffix) {
+  zval* value = nr_php_get_constant(constant);
 
   if (nr_php_is_zval_valid_string(value)) {
-    nrl_verbosedebug(NRL_FRAMEWORK,
-                     "Wordpress: found value = %s for constant=%s",
-                     Z_STRVAL_P(value), constant);
-
-    nrbuf_t* buf = nr_buffer_create(2 * Z_STRLEN_P(value), 0);
-
-    nr_regex_add_quoted_to_buffer(buf, Z_STRVAL_P(value),
-                                  zval_len_without_trailing_slash(value));
-    nr_buffer_add(buf, suffix, nr_strlen(suffix));
-    add_wildcard_path_component(buf);
-    nr_buffer_add(buf, NR_PSTR("\0"));
+    nr_matcher_t* matcher = nr_matcher_create();
+    char* prefix = nr_formatf("%s%s", Z_STRVAL_P(value), suffix);
 
-    regex = nr_regex_create(nr_buffer_cptr(buf), NR_REGEX_CASELESS, 1);
+    nr_matcher_add_prefix(matcher, prefix);
 
-    nr_buffer_destroy(&buf);
-  } else {
-    /*
-     * If the constant isn't set, that's not a problem, but if it is and it's
-     * an unexpected type we should log a message.
-     */
-    if (value) {
-      nrl_verbosedebug(NRL_FRAMEWORK, "%s: unexpected non-string value for %s",
-                       __func__, constant);
-    }
+    nr_free(prefix);
+    nr_php_zval_free(&value);
+    return matcher;
   }
 
   nr_php_zval_free(&value);
-  return regex;
+  return NULL;
 }
 
-static char* try_match_regex(const nr_regex_t* regex, const char* filename) {
-  char* plugin = NULL;
-  nr_regex_substrings_t* ss = NULL;
-
-  ss = nr_regex_match_capture(regex, filename, nr_strlen(filename));
-  if (NULL == ss) {
+static char* nr_wordpress_strip_php_suffix(char* filename) {
+  char* retval = NULL;
+  char* ext_offset_ptr = NULL;
+  size_t offset = 0;
+  if (NULL == filename) {
     return NULL;
   }
-
-  /*
-   * The last substring will be the plugin or theme name.
-   */
-  plugin = nr_regex_substrings_get(ss, nr_regex_substrings_count(ss));
-  nr_regex_substrings_destroy(&ss);
-
-  return plugin;
+  ext_offset_ptr = nr_strstr(filename, ".php");
+  if (NULL == ext_offset_ptr) {
+    // If we don't catch a .php extension, just return original
+    return filename;
+  }
+  offset = nr_strlen(filename) - nr_strlen(ext_offset_ptr);
```

Comment:
If performance improvements are what this PR is trying to achieve, then using `nr_file_basename` is not the right approach - it not only calls `nr_strlen` but also calls `nr_strdup`. I think `nr_wordpress_strip_php_suffix` should be left in this file but refactored to what I suggested in my previous comment. If `nr_wordpress_strip_php_suffix` is to become utility function, then I think to make it most efficient it should become:
- static inline to avoid function call
- accept string lengths because it is sometimes caller knows them up front (which is the case here, both the filename length and extension length are known up front) and there's no need to calculate them with `nr_strlen`

Reply:
I'm referring to this comment: https://github.com/newrelic/newrelic-php-agent/pull/765#discussion_r1401304694

Reply:
A few things regarding this suggestion:

1. Making the function static inline in the `util_strings.h` file will require importing `util_memory.h` to allow access to the `nr_strdup` and `nr_free` family of functions.
2. We will still have to call `nr_strlen` in order to get the string length of the filename in the calling function- the `filename_len` referred to in the function [nr_wordpress_plugin_from_function](https://github.com/newrelic/newrelic-php-agent/pull/765/files#diff-42efe584450b1fe50cb4d486c49549992ad10f15774d7f0b8f02d02c9f4ba2aeL291) is not the equivalent to the length of the [plugin](https://github.com/newrelic/newrelic-php-agent/pull/765/files#diff-42efe584450b1fe50cb4d486c49549992ad10f15774d7f0b8f02d02c9f4ba2aeR227) value returned from the matcher function, which is what gets passed to `nr_file_basename`. The reason for this is that [nr_matcher_match](https://github.com/newrelic/newrelic-php-agent/pull/765/files#diff-6e4026af6152229c92ebb40777f8ee44ef666b8c315d7f1eb92b63f2e512671dR58) is returning a [substring](https://github.com/newrelic/newrelic-php-agent/pull/765/files#diff-6e4026af6152229c92ebb40777f8ee44ef666b8c315d7f1eb92b63f2e512671dR92-R94) of the full, raw filename, and so another call to `nr_strlen` is required somewhere to get the value of the returned substring.

Reply:
Then it looks like `nr_wordpress_strip_php_suffix` function is too specialized and does not belong to axiom.

Reply:
Looking at the bigger picture it looks like the matcher itself could be improved a bit to avoid multiple calculations of string lengths and have it return the length of the match as it calculates this already:
1. Change the type of an element stored in prefixes vector from a string to a tuple: string + string's length; the length is already calculated by nr_matcher_add_prefix.
2. Change the return type of nr_matcher_match_* from a string (`char *`) to a tuple: string + string's length; the length of a match is calculated by `nr_matcher_match_internal`.
This optimization will avoid re-calculating matcher's prefix's string length every time the matcher is used. It will also allow for `nr_wordpress_strip_php_suffix` to use pre-calculated length of the plugin.

Reply:
Took a look at #789 and it looks reasonable. Let me resolve the outstanding feedback on this PR so it's in a "clean" state for your PR.

Reply:
Mind you that #789 is a draft that needs at least a few NULL checks for results returned by memory allocator. I'm going to close some of the outdated conversations.

Reply:
With #789 do we still want to move `nr_wordpress_strip_php_suffix` back into the fw_wordpress.c, or keep it as a utility in util_strings.h?

Reply:
It is a very specialized function so I'd move it back to fw_wordpress.c

Reply:
I think it's generalized enough to keep as a utility in case we need similar functionality in the future- there's nothing necessarily specific to wordpress inside the function. I pushed fixes to get it to work with my current PR without too much disruption to your PR in 52ce2253c518156ecd113bb9158e9d746a17e29d.
If we want to continue tweaking this function or changing where it's declared I think we should do it in #789 to avoid unnecessary conflicts

Reply:
> Mind you that https://github.com/newrelic/newrelic-php-agent/pull/789 is a draft that needs at least a few NULL checks for results returned by memory allocator. I'm going to close some of the outdated conversations.

It's no longer the case - I've added NULL checks in https://github.com/newrelic/newrelic-php-agent/pull/789.

Reply:
@bduranleau-nr #789 includes all my suggestions and is ready for review and merge.